### PR TITLE
TF-80 Move email to mailbox

### PR DESCRIPTION
**Pull request by dab246**
**== MOBILE ==**

https://user-images.githubusercontent.com/80730648/134138261-bf2a6a0f-4886-44fb-b66e-3566e021cdff.mov

**= = TABLET ==**

https://user-images.githubusercontent.com/80730648/134145137-1498949b-6852-48e7-a877-50757977f2bf.mov

### CODE REVIEW

**Comment by hoangdat** [comment]
please share the video from tablet also, how about the destination picker in tablet :)

**Comment by chibenwa** [comment]
I do not succeed to see the Video - looks broken. Can you try re-iuploading it ?

**Comment by hoangdat** [comment]
> I do not succeed to see the Video - looks broken. Can you try re-uploading it ?

I have the same prb, but it work in my email client :)

image

**Comment by dab246** [comment]
Updated video demo !!

**Comment by chibenwa** [comment]
Once again, good work! The ergonomy sounds good to me for mobile usage. Good initiative with the UNDO button. I anm eager to try it.

Here are few questions that I would like to discuss:

 - can we expend to see and select child mailboxes in the mailbox picker?

 - I would suspect the `mailbox picker` to request yet-again the full mailbox list from the backend server (there is a loading latency) while we already have that as part of the `mailboxes view`. To be honnest we should minimise the requests to the server (**especially for the list of mailboxes**) so we should not do any backend request as part of the mailbox picker. That's OK to solve this once we have the cache.
 
  - Idem moving an item should not reload the full `thread view`. Ok to solve this once we have a cache...
  
We need to keep in mind that we target million+ deployments and thus (mobile) clients should have exemplary behaviour to reduce (backend) costs to the maximum in order for TMail solutions to be competitive. We can not afford any duplicated JMAP queries... Also mailboxes are critical as some dudes gets 1000+ mailboxes and getting them is expensive!

**Comment by dab246** [comment]
> Please provide demo video to select destination in sub-folder and in folder which have sub-folder

https://user-images.githubusercontent.com/80730648/134272612-c8321ea0-4c2a-4ed3-89df-986d8c9aa5c6.mov

**Comment by hoangdat** [comment]
> folder which have sub-folder

how about we select `f5`?

**Comment by chibenwa** [comment]
>     Please provide demo video to select destination in sub-folder and in folder which have sub-folder

And moving a mail in `f5/ok` is ok? (mailbox with child folders)

**Comment by dab246** [comment]
> > folder which have sub-folder
> 
> how about we select `f5`?

I have updated the tree widget again. So it works fine.

https://user-images.githubusercontent.com/80730648/134326846-378fda7c-2e5f-48fc-b945-2170330dd237.mov

**Comment by dab246** [comment]
> > ```
> > Please provide demo video to select destination in sub-folder and in folder which have sub-folder
> > ```
> 
> And moving a mail in `f5/ok` is ok? (mailbox with child folders)

Please see the demo above

File: lib/features/base/base_controller.dart [hunk 1]
```diff
@@ -14,6 +14,7 @@ abstract class BaseController extends GetxController {
   }
 
   void dispatchState(Either<Failure, Success> newState) {
+    viewState.firstRebuild = true;
```
Comment (hoangdat) [inline lib/features/base/base_controller.dart:17]: what is it?

File: lib/features/destination_picker/presentation/destination_picker_controller.dart [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import 'package:tmail_ui_user/features/mailbox/domain/state/get_all_mailboxes_state.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/usecases/get_all_mailbox_interactor.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/model/mailbox_node.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/model/mailbox_tree.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/model/mailbox_tree_builder.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/extensions/list_mailbox_node_extension.dart';
+import 'package:get/get.dart';
+import 'package:tmail_ui_user/main/routes/route_navigation.dart';
+
+class DestinationPickerController extends BaseController {
```
Comment (hoangdat) [inline lib/features/destination_picker/presentation/destination_picker_controller.dart:16]: please check the comment: https://github.com/linagora/tmail-flutter/pull/115#issuecomment-924124966. For this picker, we don't need to geed it from the server.

Comment (hoangdat) [reply lib/features/destination_picker/presentation/destination_picker_controller.dart:16]: Please set up the destination picker with the model: `in/out`
```
mailboxes -> user interaction in destination picker -> done button -> the picked mailbox send back to the invoker -> mailbox dashboard will execute the moving task -> toast
```
with this, we can reuse the destination picker in many places

File: lib/features/mailbox/presentation/mailbox_view.dart [hunk 3]
```diff
@@ -189,30 +188,38 @@ class MailboxView extends GetWidget<MailboxController> {
...
+                  isExpanded: mailboxNode.expandMode == ExpandMode.EXPAND,
+                  parent: Obx(() => (MailBoxFolderTileBuilder(
+                          context,
+                          imagePaths,
+                          responsiveUtils,
+                          mailboxNode,
+                          selectMode: controller.getSelectMode(
+                              mailboxNode.item,
+                              controller.mailboxDashBoardController.selectedMailbox.value))
+                      ..addOnSelectMailboxFolderClick((mailboxNode) => controller.selectMailbox(context, mailboxNode.item))
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_view.dart:205]: cool

File: lib/features/base/base_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/base/base_controller.dart:17]: set a flag to indicate this value is heard

File: lib/features/destination_picker/presentation/destination_picker_controller.dart (hunk 2 above)
Comment (dab246) [reply lib/features/destination_picker/presentation/destination_picker_controller.dart:16]: I don't understand what you say. Call me to speak more clearly.

File: lib/features/thread/domain/usecases/move_multiple_email_to_mailbox_interactor.dart [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import 'package:dartz/dartz.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/mail/email/email.dart';
+import 'package:model/email/move_action.dart';
+import 'package:model/mailbox/presentation_mailbox.dart';
+import 'package:tmail_ui_user/features/email/domain/model/move_request.dart';
+import 'package:tmail_ui_user/features/email/domain/usecases/move_to_mailbox_interactor.dart';
+import 'package:tmail_ui_user/features/thread/domain/state/move_multiple_email_to_mailbox_state.dart';
+
+class MoveMultipleEmailToMailboxInteractor {
```
Comment (hoangdat) [inline lib/features/thread/domain/usecases/move_multiple_email_to_mailbox_interactor.dart:11]: class MoveMultipleEmailToMailboxInteractor {
  final MoveToMailboxInteractor _moveToMailboxInteractor;

  MoveMultipleEmailToMailboxInteractor(this._moveToMailboxInteractor);

  Stream<Either<Failure, Success>> execute(
      AccountId accountId,
      List<EmailId> listEmailId,
      PresentationMailbox currentMailbox,
      PresentationMailbox destinationMailbox,
      MoveAction moveAction,
  ) async* {
    try {
      yield Right<Failure, Success>(UIState.loading);
      
      final streamBuilder = StreamController<Either<Failure, Success>>();
      
      listEmailId.map((emailId) => streamBuilder.addStream(
        _moveToMailboxInteractor.execute(
          accountId,
          MoveRequest(emailId, currentMailbox, destinationMailbox, moveAction)))
      );

      final listResult = await streamBuilder.stream
        .where(_validateSuccessState)
        .toList();
      
      streamBuilder.close();
      
      if (listEmailId.length == listResult.length) {
        yield Right(MoveMultipleEmailToMailboxAllSuccess(listEmailId, currentMailbox, destinationMailbox, moveAction));
      } else if (listResult.length == 0) {
        yield Left(MoveMultipleEmailToMailboxAllFailure(moveAction));
      }
      yield Right(MoveMultipleEmailToMailboxHasSomeEmailFailure(listEmailId, currentMailbox, destinationMailbox, moveAction));
    } catch (e) {
      yield Left(MoveMultipleEmailToMailboxFailure(e, moveAction));
    }
  }

  bool _validateSuccessState(Either<Failure, Success> state) {
    if (state.isRight() && state is UIState) {
      if (state is MoveToMailboxSuccess) {
        return true;
      }
    }
    return false;
  }
}

Comment (dab246) [reply lib/features/thread/domain/usecases/move_multiple_email_to_mailbox_interactor.dart:11]: Please don't make miscellaneous comments here. Please give a possible solution. Be a worthy leader.

Comment (hoangdat) [reply lib/features/thread/domain/usecases/move_multiple_email_to_mailbox_interactor.dart:11]: :), please pair it tmr. We need to find the better solution for it.

Comment (hoangdat) [reply lib/features/thread/domain/usecases/move_multiple_email_to_mailbox_interactor.dart:11]: sorry :), I come back with better version: need to use `async` package:

```dart
import 'dart:async';

import 'package:async/async.dart';
import 'package:core/core.dart';
import 'package:dartz/dartz.dart';
import 'package:jmap_dart_client/jmap/account_id.dart';
import 'package:jmap_dart_client/jmap/mail/email/email.dart';
import 'package:model/email/move_action.dart';
import 'package:model/mailbox/presentation_mailbox.dart';
import 'package:tmail_ui_user/features/email/domain/model/move_request.dart';
import 'package:tmail_ui_user/features/email/domain/state/move_to_mailbox_state.dart';
import 'package:tmail_ui_user/features/email/domain/usecases/move_to_mailbox_interactor.dart';
import 'package:tmail_ui_user/features/thread/domain/state/move_multiple_email_to_mailbox_state.dart';

class MoveMultipleEmailToMailboxInteractor {
  final MoveToMailboxInteractor _moveToMailboxInteractor;

  MoveMultipleEmailToMailboxInteractor(this._moveToMailboxInteractor);

  Stream<Either<Failure, Success>> execute(
      AccountId accountId,
      List<EmailId> listEmailId,
      PresentationMailbox currentMailbox,
      PresentationMailbox destinationMailbox,
      MoveAction moveAction,
  ) async* {
    try {
      yield Right<Failure, Success>(UIState.loading);
      
      final mergedStream = StreamGroup.merge(
        listEmailId.map((emailId) => _moveToMailboxInteractor.execute(
            accountId,
            MoveRequest(emailId, currentMailbox, destinationMailbox, moveAction))
      ));
      
      final listResult = await mergedStream
        .where(_validateSuccessState)
        .toList();
      
      if (listEmailId.length == listResult.length) {
        yield Right(MoveMultipleEmailToMailboxAllSuccess(listEmailId, currentMailbox, destinationMailbox, moveAction));
      } else if (listResult.length == 0) {
        yield Left(MoveMultipleEmailToMailboxAllFailure(moveAction));
      }
      yield Right(MoveMultipleEmailToMailboxHasSomeEmailFailure(listEmailId, currentMailbox, destinationMailbox, moveAction));
    } catch (e) {
      yield Left(MoveMultipleEmailToMailboxFailure(e, moveAction));
    }
  }

  bool _validateSuccessState(Either<Failure, Success> state) {
    if (state.isRight() && state is UIState) {
      if (state is MoveToMailboxSuccess) {
        return true;
      }
    }
    return false;
  }
}
```

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 5]
```diff
@@ -26,8 +28,9 @@ class MailboxController extends BaseController {
   final TreeBuilder _treeBuilder;
   final ResponsiveUtils responsiveUtils;
 
-  final folderMailboxTree = MailboxTree(MailboxNode.root()).obs;
+  MailboxTree folderMailboxTree = MailboxTree(MailboxNode.root());
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:31]: It seems we don't need this field.

File: lib/features/base/base_controller.dart (hunk 1 above)
Comment (hoangdat) [reply lib/features/base/base_controller.dart]: so with StreamZip, we don't need this change, am I right?

Comment (dab246) [reply lib/features/base/base_controller.dart]: yes.

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 5 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:31]: yes.

File: lib/features/email/domain/repository/email_repository.dart [hunk 6]
```diff
@@ -3,13 +3,14 @@ import 'package:jmap_dart_client/jmap/account_id.dart';
...
 import 'package:tmail_ui_user/features/composer/domain/model/email_request.dart';
+import 'package:tmail_ui_user/features/email/domain/model/move_request.dart';
 
 abstract class EmailRepository {
   Future<Email> getEmailContent(AccountId accountId, EmailId emailId);
 
   Future<bool> sendEmail(AccountId accountId, EmailRequest emailRequest);
 
-  Future<bool> markAsRead(AccountId accountId, EmailId emailId, ReadActions readAction);
+  Future<List<Email>> markAsRead(AccountId accountId, List<PresentationEmail> presentationEmails, ReadActions readActions);
```
Comment (hoangdat) [inline lib/features/email/domain/repository/email_repository.dart:13]: Should be 
```dart
Future<List<Email>> markAsRead(AccountId accountId, List<Email> emails, ReadActions readActions);
```

File: lib/features/email/domain/repository/email_repository.dart [hunk 7]
```diff
@@ -25,4 +26,6 @@ abstract class EmailRepository {
     AccountRequest accountRequest,
     CancelToken cancelToken
   );
+
+  Future<List<EmailId>> moveToMailbox(AccountId accountId, MoveRequest moveRequest);
```
Comment (hoangdat) [inline lib/features/email/domain/repository/email_repository.dart:30]: same here.

**Review by hoangdat — APPROVED** [review]

**Pull request merged**